### ISSUE TITLE
freeze discriminator and train generator only

### ISSUE THREAD

**Issue by pretidav** [google-research/electra#37]
Hello, 

I was wondering about the possibility of making un-trainable the discriminator for the first n steps, to give time to generator to learn how to produce meaningful tokens before feeding them to the discriminator. 
Since the the training of generator and discriminator are not adversarial (as far as I understand from the paper) this should be possible. 
Is this option already implemented and I am missing some flag? 
Would be difficult to implement it? Maybe some workaround with existent parameters? 

Moreover, I see that the global loss that's a linear combination of the generator and the discriminator 
appear by default heavily unbalance respect to discriminator (with a relative weight of 50), this of course make sense since the "real" model is the discriminator. 
Do you believe that re-weighting the losses can enhance the generator training?  

Thank you very much for the answers.

**Comment by clarkkev** [comment]
Hi,

For your first question, we never did this because the generator pretty quickly learns to produce meaningful tokens. In fact, we more often ran into the exact opposite issue where if the generator gets too good too quickly the discriminator fails to learn anything because the real/fake task is too hard. The easiest way of implementing this would be multiplying the discriminator loss by something that is 0 for the first `n` steps, e.g., `self.total_loss += config.disc_weight * tf.cast(tf.greater_equal(tf.train.get_or_create_global_step(), 1000), tf.float32) * disc_output.loss`.

For your second question: yes the weighting is making the generator a bit worse and the discriminator a bit better because at the end of the day we care about the discriminator for downstream tasks. One other thing to note is that the loss weights only matters when there are tied parameters between the discriminator and generator (the token embeddings are shared by default). Otherwise there is no "competition" between the generator/discriminator tasks and scaling the loss doesn't matter since we use Adam.

**Comment by pretidav** [comment]
Thank you very much for this answer!
